client_api/api/server.go

```diff
@@ -0,0 +1,799 @@
+package api
+
+import (
+	"path/filepath"
+
+	"github.com/Sirupsen/logrus"
+	google_protobuf "github.com/golang/protobuf/ptypes/empty"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"crypto/rand"
+	"github.com/docker/notary"
+	"github.com/docker/notary/client"
+	"github.com/docker/notary/client/changelist"
+	"github.com/docker/notary/cryptoservice"
+	"github.com/docker/notary/storage"
+	"github.com/docker/notary/trustmanager"
+	"github.com/docker/notary/trustpinning"
+	"github.com/docker/notary/tuf/data"
+	"github.com/docker/notary/utils"
+)
+
+// NewServer creates a new instance of a Client API server with a configured
+// upstream Notary Server.
+func NewServer(upstream string, upstreamCAPath string, grpcSrv *grpc.Server) error {
+	srv := &Server{
+		upstream:       upstream,
+		upstreamCAPath: upstreamCAPath,
+	}
+	RegisterNotaryServer(grpcSrv, srv)
+	return nil
+}
+
+type Server struct {
+	upstream       string
+	upstreamCAPath string
+}
+
+func (srv *Server) Initialize(ctx context.Context, initMessage *InitMessage) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(initMessage.Gun))
+	if err != nil {
+		return nil, err
+	}
+
+	roles := make([]data.RoleName, len(initMessage.ServerManagedRoles.Roles))
+	for index, role := range initMessage.ServerManagedRoles.Roles {
+		roles[index] = data.RoleName(role)
+	}
+
+	err = r.Initialize(initMessage.RootKeyIDs, roles...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) Publish(ctx context.Context, gun *GunMessage) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(gun.Gun))
+	if err != nil {
+		return nil, err
+	}
+
+	err = r.Publish()
+	if err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) AddTarget(ctx context.Context, t *Target) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(t.GetGun()))
+	if err != nil {
+		return nil, err
+	}
+	if err := r.AddTarget(
+		&client.Target{
+			Name:   t.GetName(),
+			Hashes: data.Hashes(t.Hashes),
+			Length: t.Length,
+		},
+	); err != nil {
+		return nil, err
+	}
+	if err := publishRepo(r); err != nil {
+		return nil, err
+	}
+
+	return &BasicResponse{
+		Success: true,
+	}, nil
+}
+
+func (srv *Server) RemoveTarget(ctx context.Context, t *Target) (*BasicResponse, error) {
+	r, err := srv.initRepo(data.GUN(t.GetGun()))
+	if err != nil {
+		return nil, err
+	}
+	if err := r.RemoveTarget(
+		t.GetName(), "targets",
+	); err != nil {
+		return nil, err
+	}
+	if err := publishRepo(r); err != nil {
```

Comment:
Should the responsibility to publish be pushed back to the "clientapi client" instead of in the "clientapi server"?

IIRC this mirrors the current cli behavior, just wondering if there's a good reason to keep them coupled together.